src/core/instances.jl

```diff
@@ -0,0 +1,314 @@
+value(param::ScalarModelParameter) = param.value
+
+value(param::ArrayModelParameter)  = param.values
+
+dimensions(obj::ArrayModelParameter) = obj.dimensions
+
+dimensions(obj::ScalarModelParameter) = []
+
+"""
+    variables(mi::ModelInstance, comp_name::Symbol)
+
+Return the `ComponentInstanceVariables` for `comp_name` in ModelInstance 'mi'.
+"""
+variables(mi::ModelInstance, comp_name::Symbol) = variables(compinstance(mi, comp_name))
+
+variables(ci::ComponentInstance) = ci.variables
+
+"""
+    parameters(mi::ModelInstance, comp_name::Symbol)
+
+Return the `ComponentInstanceParameters` for `comp_name` in ModelInstance 'mi'.
+"""
+parameters(mi::ModelInstance, comp_name::Symbol) = parameters(compinstance(mi, comp_name))
+
+parameters(ci::ComponentInstance) = ci.parameters
+
+
+function getindex(mi::ModelInstance, comp_name::Symbol, datum_name::Symbol)
+    if !(comp_name in keys(mi.components))
+        error("Component does not exist in current model")
+    end
+    
+    comp_inst = compinstance(mi, comp_name)
+    vars = comp_inst.variables
+    pars = comp_inst.parameters
+
+    if datum_name in vars.names
+        which = vars
+    elseif datum_name in pars.names
+        which = pars
+    else
+        error("$datum_name is not a parameter or a variable in component $comp_name.")
+    end
+
+    value = getproperty(which, Val(datum_name))
+    # return isa(value, PklVector) || isa(value, TimestepMatrix) ? value.data : value
+    return isa(value, AbstractTimestepMatrix) ? value.data : value
+end
+
+# """
+#     indexcount(mi::ModelInstance, idx_name::Symbol)
+
+# Returns the size of index `idx_name`` in model instance `mi`.
+# """
+# indexcount(mi::ModelInstance, idx_name::Symbol) = indexcounts(mi.md, idx_name)
+
+"""
+    dim_count(mi::ModelInstance, dim_name::Symbol)
+
+Returns the size of index `dim_name`` in model instance `mi`.
+"""
+dim_count(mi::ModelInstance, dim_name::Symbol) = dim_count(mi.md, dim_name)
+
+dim_key_dict(mi::ModelInstance) = dim_key_dict(mi.md)
+
+dim_value_dict(mi::ModelInstance) = dim_value_dict(mi.md)
+
+# """
+#     indexvalues(mi::ModelInstance, idx_name::Symbol)
+
+# Return the values of index `idx_name` in model instance `mi`.
+# """
+# indexvalues(mi::ModelInstance, idx_name::Symbol) = indexvalues(mi.md, idx_name)
+
+function make_clock(mi::ModelInstance, ntimesteps, time_keys::Vector{Int})
+    start = time_keys[1]
+    stop  = time_keys[min(length(time_keys), ntimesteps)]
+    step  = step_size(time_keys)
+    return Clock(start, step, stop)
+end
+
+function reset_variables(ci::ComponentInstance)
+    # println("reset_variables($(ci.comp_id))")
+    vars = ci.variables
+
+    for (name, ref) in zip(vars.names, vars.types.parameters)
+        # Everything is held in a Ref{}, so get the parameters to that...
+        T = ref.parameters[1]
+        value = getproperty(vars, Val(name))
+
+        if (T <: AbstractTimestepMatrix || T <: AbstractMatrix) && eltype(value) <: AbstractFloat
```

Comment:
Shouldn't this be the case for any ``AbstractArray``?

Reply:
Sure. I made `AbstractTimestepMatrix <: AbstractArray` and changed it to test for that.